File: LeetCode/Array/948. Bag of Tokens.java
/*
    @kartik chauhan
    Leetcode : https://leetcode.com/problems/bag-of-tokens/
*/
import java.util.*;

class Solution {
    public int bagOfTokensScore(int[] tokens, int power) {
        
        int res = 0;
        Arrays.sort(tokens);
        int i = 0, j = tokens.length - 1;
        int p = 0;

        while(i <= j && (power >= tokens[i] || p > 0)) {
            while(i <=j && power >= tokens[i]) {
                power -= tokens[i++];
                p += 1;
            }

            res = Math.max(res, p);

            if(i <= j && p > 0) {
                power += tokens[j--];
                p -= 1;
            }
        }
        return res;
    }
}

File: LeetCode/Array/948. Bag of Tokens.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    int bagOfTokensScore(vector<int>& tokens, int power) {
        int res = 0;
        sort(tokens.begin(), tokens.end());
        int i = 0, j = tokens.size() - 1;
        bool flag = true;
        while (i <= j && flag)
        {
            if(power >= tokens[i]) {
                power -= tokens[i];
                i += 1;
                res += 1;
            }    
            else if(res >= 1) {
                if(power + tokens[j] >= tokens[i] and i != j) {
                    power += tokens[j];
                    j -= 1;
                    res -= 1;
                }
                else {
                    flag = false;
                }
            }
            else {
                break;
            }
        }
        return res;
    }
};